truss/templates/trtllm/packages/client.py

```diff
@@ -0,0 +1,159 @@
+import json
+import os
+import subprocess
+import time
+from pathlib import Path
+from typing import AsyncGenerator, Optional
+
+import numpy as np
+import tritonclient.grpc.aio as grpcclient
+import tritonclient.http as httpclient
+from utils import (
+    GRPC_SERVICE_PORT,
+    HTTP_SERVICE_PORT,
+    is_triton_server_alive,
+    move_data_to_model_repository,
+    prepare_grpc_tensor,
+)
+
+
+class TritonClient:
+    def __init__(self, data_dir: Path, model_repository_dir: Path, parallel_count=1):
+        self._data_dir: Path = data_dir
+        self._model_repository_dir: Path = model_repository_dir
+        self._parallel_count = parallel_count
+        self._http_client: Optional[httpclient.InferenceServerClient] = None
+        self.grpc_client_instance: Optional[grpcclient.InferenceServerClient] = None
+
+    def start_grpc_stream(self) -> grpcclient.InferenceServerClient:
+        if self.grpc_client_instance:
+            return self.grpc_client_instance
+
+        return grpcclient.InferenceServerClient(
+            url=f"localhost:{GRPC_SERVICE_PORT}", verbose=False
+        )
+
+    def load_server_and_model(self, env: dict) -> None:
+        """Loads the Triton server and the model."""
+        if not is_triton_server_alive():
+            move_data_to_model_repository(self._data_dir)
+            self.start_server(mpi=self._parallel_count, env=env)
+
+        self._http_client = httpclient.InferenceServerClient(
+            url=f"localhost:{HTTP_SERVICE_PORT}", verbose=False
+        )
+
+        is_server_up = False
+        while not is_server_up:
+            try:
+                is_server_up = self._http_client.is_server_live()
+            except ConnectionRefusedError:
+                time.sleep(2)
+                continue
+
+        while not self._http_client.is_model_ready(model_name="ensemble"):
+            time.sleep(2)
```

Comment:
I just happened to got hung up with this while playing with spec dec which does not have an ensemble. Can me find a condition that is not depending on a specific model name? Maybe `is_server_ready()`?